File: ten-basic-algorithms/capitalizeFirst/capitalizeFirst.ts
function capitalizeFirstLetter(str: string): string {
    // Split the string into words, capitalize the first letter of each word, and join them back together
    // using a space as the separator.
    const words = str.split(' ');
    // Use map to iterate over each word, capitalize the first letter, and join them back together
    return words.map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
}

capitalizeFirstLetter('look, it is working!');

File: basic-algorithms/capitalizeFirst/capitalizeFirst.go
package main

import(
	"unicode"
	"strings"
	"fmt"
)

func capitalizeFirstLetter(str string) string {
	// Split the string into words
	// and capitalize the first letter of each word
	words := strings.Split(str, " ")
	// Iterate over each word and capitalize the first letter
	// using unicode.ToUpper to handle Unicode characters
	for i, word := range words {
		if len(word) > 0 {
			r := []rune(word)
			r[0] = unicode.ToUpper(r[0])
			words[i] = string(r)
		}
	}
	// Join the words back into a single string
	// and return the result
	return strings.Join(words, " ")
}

func main() {
	str := "look, it is working!"
	capitalizedStr := capitalizeFirstLetter(str)
	fmt.Println(capitalizedStr)
}